### PR TITLE
Add tab completion for newer IPython.

### LINKED ISSUE #819: Tab completion broken in newer IPython versions
In recent IPython versions, tab completion has changed. cocotb uses ``_getAttributeNames`` to allow completing DUT object names but this has been removed in https://github.com/ipython/ipython/pull/8081/files.

These days, providing ``__dir__(self)`` is the way to go (https://ipython.readthedocs.io/en/stable/config/integrating.html). I'll prepare a PR.

### PR DESCRIPTION
**Pull request by cmarqu**
Fixes #819.
Add test showing ``IPython.embed_kernel()`` because only this handles keystrokes correctly for me.

### CODE REVIEW

File: cocotb/handle.py [hunk 1]
```diff
@@ -216,10 +216,15 @@ def _sub_handle_key(self, name):
         return name.split(".")[-1]
 
     def _getAttributeNames(self):
-        """Permits IPython tab completion to work."""
+        """Permits older IPython tab completion to work."""
         self._discover_all()
```
Comment (eric-wieser) [inline cocotb/handle.py]: You can remove this line now, since `dir` handles it

Comment (eric-wieser) [reply cocotb/handle.py]: Actually, I'd just remove this function altogether - `__dir__` has been supported by ipython from its very first git commit:

https://github.com/takluyver/ipython/blame/11c30a1d83c24ccad09f8f8063a5f8ba36c3137d/IPython/genutils.py#L1924-L1954

File: cocotb/handle.py [hunk 2]
```diff
@@ -216,10 +216,15 @@ def _sub_handle_key(self, name):
...
         return dir(self)
 
+    def __dir__(self):
+        """Permits newer IPython tab completion to work."""
+        self._discover_all()
+        return super().__dir__() + [str(k) for k in self._sub_handles]
```
Comment (eric-wieser) [inline cocotb/handle.py]: This won't work in python 2 - needs to be `super(RegionObject, self).__dir__()`

Comment (cmarqu) [reply cocotb/handle.py]: Thanks, fixed.

File: cocotb/handle.py (hunk 1 above)
Comment (cmarqu) [reply cocotb/handle.py]: Removed.

File: tests/test_cases/test_discovery/test_discovery.py [hunk 3]
```diff
@@ -199,35 +219,35 @@ def access_string(dut):
     test_string = "cocotb"
     dut.stream_in_string.setimmediatevalue(test_string)
 
-    variable_string = dut.stream_out_string
```
Comment (garmin-mjames) [inline tests/test_cases/test_discovery/test_discovery.py]: Was this change intentional? "Variable" is the correct spelling.

Comment (cmarqu) [reply tests/test_cases/test_discovery/test_discovery.py]: Oh, thanks, I somehow reverted a change of my own from another PR. I'll rework this next week.

**Comment by imphil** [comment]
LGTM, thanks!
